File: 2025/day5/task1.py
import time


def input_parser(filename: str):
    ranges = []
    ids = []
    with open(filename, "r") as file:
        data = file.read().strip().split("\n")
        split_index = data.index("")
        ranges = list(
            map(lambda range: list(map(int, range.split("-"))), data[:split_index])
        )
        ids = list(map(int, data[split_index + 1 :]))
    return ranges, ids


def resolve_range(ranges: list[list]) -> list[list]:
    ranges.sort(key=lambda x: x[0])
    stack = []
    for range in ranges:
        if not stack:
            stack.append(range)
            continue
        curr_start, curr_end = range
        prev_start, prev_end = stack.pop()
        if curr_start > prev_end:
            stack.append([prev_start, prev_end])
            stack.append([curr_start, curr_end])
        else:
            stack.append([prev_start, max(curr_end, prev_end)])
    return stack


def counter(ranges: list[list], ids: list[int]) -> int:
    ids.sort()
    i = 0
    count = 0
    for id in ids:
        while i < len(ranges):
            if ranges[i][0] <= id <= ranges[i][1]:
                count += 1
                break
            if ranges[i][0] > id:
                break
            if ranges[i][1] < id:
                i += 1
    return count


if __name__ == "__main__":
    start_time = time.time()
    ranges, ids = input_parser("data.txt")
    ranges = resolve_range(ranges)
    count = counter(ranges, ids)
    print(count)
    end_time = time.time()
    print(f"Execution time: {end_time - start_time:.6f} seconds")

File: 2025/day5/task1.cpp
#include <stdio.h>
#include <vector>
#include <string>
#include <fstream>
#include <iostream>

struct Range{
    long long start;
    long long end;
};

std::pair<std::vector<Range>, std::vector<long>> input_parser(std::string filename){
    std::ifstream file{std::string(filename)};
    std::vector<Range> ranges;
    std::vector<long> ids;

    bool readRange = true;
    std::string line;
    while (std::getline(file, line)) {
        if(line.empty()){
            readRange = false;
            continue;
        }
        if(readRange){
            int start, end;
            Range range;
            range.start = std::stoll(line.substr(0, line.find("-")));;
            range.end = std::stoll(line.substr(line.find("-")+1, line.size()));;
            ranges.push_back(range);
        }
        else{
            ids.push_back(std::stoll(line));
        }
    }
    return {ranges, ids};
}

std::vector<Range> resolve_ranges(std::vector<Range> ranges){
    std::sort(ranges.begin(), ranges.end(), [](Range& a, Range& b){
        return a.start < b.start;
    });

    std::vector<Range> new_ranges;
    for(int i = 0; i < ranges.size(); i++){
        if(new_ranges.size() == 0){
            new_ranges.push_back(ranges[i]);
            continue;
        }
        Range& prev = new_ranges.back();
        if(ranges[i].start > prev.end){
            new_ranges.push_back(ranges[i]);
        }
        else{
            prev.end = std::max(ranges[i].end, prev.end);
        }
    }
    return new_ranges;
}


int counter(std::vector<Range> ranges, std::vector<long> ids){
    int i = 0;
    int fresh = 0;
    for(long& id: ids){
        while(i < ranges.size()){
            if(ranges[i].start <= id && id <= ranges[i].end){
                fresh += 1;
                break;
            }
            if(ranges[i].start > id){
                break;
            }
            if(ranges[i].end < id){
                i += 1;
            }
        }
    }
    return fresh;
}


int main(){
    auto result = input_parser("data.txt");
    auto& ranges = result.first;
    auto& ids    = result.second;
    std::vector<Range> new_ranges = resolve_ranges(result.first);
    int count = counter(new_ranges,ids);
    std::cout << count << std::endl;
    return 0;
}
